kafka-connect-http/src/main/java/com/github/castorm/kafka/connect/http/HttpSourceTask.java

```diff
@@ -121,8 +128,32 @@ private HttpResponse execute(HttpRequest request) {
     }
 
     @Override
-    public void commitRecord(SourceRecord record) {
-        offset = Offset.of(record.sourceOffset());
+    public void commitRecord(SourceRecord record, RecordMetadata metadata) throws InterruptedException {
+        committedOffsets.replace(record.sourceOffset(), true);
+
+        log.debug("Committed offset {}", record.sourceOffset());
+    }
+
+    @Override
+    public void commit() throws InterruptedException {
+        Map<String, ?> newOffset = null;
+        for (Map.Entry<Map<String, ?>, Boolean> offsetEntry : committedOffsets.entrySet()) {
+            final Boolean offsetWasCommitted = offsetEntry.getValue();
+            final Map<String, ?> sourceOffset = offsetEntry.getKey();
+            if (offsetWasCommitted) {
+                newOffset = sourceOffset;
+            } else {
+                log.warn("Found uncommitted offset {}. Will resume polling from previous offset. This might result in a number of duplicated records.", sourceOffset);
+
+                break;
+            }
+        }
+
+        if (newOffset != null) {
+            offset = Offset.of(newOffset);
+        }
+
+        log.info("Offset set to {}", offset);
```

Comment:
I might be missing something but records are ordered when polled, which implies there is a sorting criteria.

That means all you should need here is the highest committed offset, right? And for that you can directly compare in `commitRecord()` the recently committed offset vs the highest committed offset so far.

This means all the state you'd need is what you already had (an `Offset` that would be the highest committed `Offset` now), instead of the full `LinkedHashMap` with all records' offsets.

Does this make sense?

Reply:
> all you should need here is the highest committed offset

It depends on the kind of delivery guarantee you want to support in case of error ("at least once" vs. "at most once"). Example of a batch of records (regardless of the order in which they were committed):

Offset | Committed | At least once | At most once
------------ | ------------- | ------------- | -------------
1 | `true` | ok | ok
2 | `false` | ok - resume polling from previous offset (1) | lost!
3 | `true` | duplicated | ok - resume polling from this offset (3)
4 | `false` | ok | ok
5 | `true` | duplicated | ok

> you can directly compare the recently committed offset vs the highest committed offset

Unless I'm missing something, comparing offsets is not that easy because they can have multiple keys that can be ASC or DESC. That's where the `LinkedHashMap` comes in handy.

Reply:
Sorry, I was making some assumptions about KC implementation, obviously my suggestion didn't make sense. My main concern was the potential for misalignment between the task state while running and the task state after a restart.

Sharing state between `poll()` and `commit()`/`commitRecord()` opens up for this misalignment, because state is no longer represented as the last committed offset, it's somehow richer, and while running we'll have more information than after a crash. Opening up for other ways of losing records.

For instance, if I'm not wrong, this scenario is possible:
**Scenario 1**
0. Records are polled
1. Offset #1 is written 
2. Record #1 is sent
3. Record #1 is not acked
4. Written offsets are committed
5. FAULT
6. Connector restarts
7. Connector restore state from offset #1
8. Offset left record #1 behind --> Record #1 lost

This is obviously not a consequence of this PR, but because of using producer acks on partitioned topics (which I reckon to be a common scenario), the only reason I mention it, is because of we are looking at the delivery guarantees, I wanted to make sure we were considering all scenarios.

Now, about the PR, if I may, I'd like to suggest one final change before merging. And I'm happy to do it myself afterwards if you don't have the time for it. Obviously, your opinion on the matter is more than welcome:

I would hide the changes in this class behind an implementation of this interface (naming also open to discussion):
```java
public interface ConfirmationWindow<T> {
    confirm(T offset);
    T getLowWatermarkOffset();
}
```

I think what you implemented is pretty much a Confirmation/Acknowledgement window, so abstracting it away would help simplifying the Task.

Thanks,
Best regards.

Reply:
Hello @castorm,

I extracted the delivery guarantee logic from the task. It's a bit rough, but I wanted to have your opinion before polishing it up, add a few tests, and maybe write some documentation...

WDYT?

Reply:
Hi @castorm,

I work with @poulinjulien.

AFAIK, `commitRecord()` is called in the callback passed to the producer. This callback is called only when the send has been ack'ed. If the send failed, or it hasn't got the time to be ack'ed when the fault occurs, the callback is not called, neither the `commitRecord`.

Also `commit()` is called at the end of the `workerSourceTask.commitOffsets()` that is always called at the end of its `execute()`. During `commitOffsets()`, the offsets that have been indeed sent (because it waits for the ongoing sending to be finished) are flushed to the `offsetStorage` (and each time a record is sent, its offset is set to the offset writer use to write on the storage). 

I think the scenario is then more likely the following:

1. Records are polled
2. Record 1 is sent
3. Record 1 is not acked
4. Written offsets are committed (i.e. none)
5. FAULT
6. Connector restarts
7. Connector restore state from offset (i.e. none or the last one committed before this scenario)
8. Records are polled

So I'm not sure the scenario you have described is possible. As the Kafka connect sequence of actions is not something straightforward, I might have missed something. What do you think?

Reply:
Hi @toch,

I might be wrong, but I think in the scenario I described only 5 and 6 aren't out of order, I'd say the rest is as I described it.

As you'll see below, `offsetWriter.offset(record.sourcePartition(), record.sourceOffset());` happens before sending the record, and in case of a fault that stops the execution of the task, offsets are committed.

I've left the relevant pieces from the [source code](https://github.com/apache/kafka/blob/77a89fcf8d7fa01831683327727f11f4bed97319/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java) below:
```java
    public void execute() {
        try {
            ...
            while (!isStopping()) {
                ...
                if (toSend == null) {
                    toSend = poll();
                    ...
                }
                if (!sendRecords())
                    stopRequestedLatch.await(SEND_FAILED_BACKOFF_MS, TimeUnit.MILLISECONDS);
            }
        } finally {
            commitOffsets();
        }
    }

    private boolean sendRecords() {
        ...
        final SourceRecordWriteCounter counter =
                toSend.size() > 0 ? new SourceRecordWriteCounter(toSend.size(), sourceTaskMetricsGroup) : null;
        for (final SourceRecord preTransformRecord : toSend) {
            ...
            synchronized (this) {
                if (!lastSendFailed) {
                    ...
                    offsetWriter.offset(record.sourcePartition(), record.sourceOffset());
                }
            }
            try {
                ...
                producer.send(
                        producerRecord,
                        new Callback() {
                            @Override
                            public void onCompletion(RecordMetadata recordMetadata, Exception e) {
                                ...
                            }
                        });
                lastSendFailed = false;
            } catch (* e) {
                ...
            }
            ...
        }
        toSend = null;
        return true;
    }

    public boolean commitOffsets() {
        ...
        Future<Void> flushFuture = offsetWriter.doFlush(new org.apache.kafka.connect.util.Callback<Void>() {
            @Override
            public void onCompletion(Throwable error, Void result) {
                ...
            }
        });
        try {
            flushFuture.get(Math.max(timeout - time.milliseconds(), 0), TimeUnit.MILLISECONDS);
            ...
        } catch (* e) {
            ...
        }
        ...
    }
```

Reply:
> Hello @castorm,
> 
> I extracted the delivery guarantee logic from the task. It's a bit rough, but I wanted to have your opinion before polishing it up, add a few tests, and maybe write some documentation...
> 
> WDYT?

Hi @poulinjulien,

To be honest, I'm not quite convinced with the current separation of concerns. `DeliveryStrategy` does take some complexity out of the task, but its definition seems driven by convenience rather than abstraction, e.g. methods don't really do what they advertise:`send()` doesn't send anything, or `commitRecord()` and `commit()` don't really commit anything anywhere.

I would encourage you to try and implement the interface I suggested or discuss an alternative before implementing it. If you don't have the time or the will, just let me know and I can try it myself.

I don't want to be delaying the PR more than necessary, so whenever you think it's ready, let me know and I will merge. We can deal with these details afterwards.

Reply:
Hi @castorm,

> As you'll see below, writing the offset (offsetWriter.offset(record.sourcePartition(), record.sourceOffset())) happens right before sending the record (producer.send(...)), and in case of a fault that stops the execution of the task, offsets are committed (commitOffsets()).

As far as I understand, the offsets kept aside by `offsetWriter.offset(...)` are not committed in case of fault because the task wait that all the sends are completed. In the synchronized bloc, the task [will loop over all the sends until they are all completed](https://github.com/apache/kafka/blob/77a89fcf8d7fa01831683327727f11f4bed97319/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L434).  If it lasts more than the given timeout, it ends [prematurely the method `commitOffsets()`](https://github.com/apache/kafka/blob/77a89fcf8d7fa01831683327727f11f4bed97319/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L441). [Same if there are any exceptions occurring during the flush](https://github.com/apache/kafka/blob/77a89fcf8d7fa01831683327727f11f4bed97319/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java#L451).

Isn't my understanding correct?

If it's correct, I don't think the faulty presented scenario is possible as neither the offsets would be committed, neither the task offsets kept in memory.

According to my experience with Kafka Connect, I haven't seen KC committing offsets for unacknowledged send.

Reply:
Hi @toch,

You are surely right. I didn't spend time enough to fully understand what's going on there. But what you pointed out makes sense.

Reply:
:bow: Thanks for the discussion. It was interesting to take the time to challenge our respective understanding of that important mechanism.

Reply:
Hi @castorm,

I made some modifications/simplifications w.r.t. your comments. Please let me know if it's ok.